File: benchmark-2/backend/backend.py
import flask
import datetime

app = flask.Flask(__name__)

@app.route("/what-date-is-it", methods=["GET"])
def dateHandler():
	date = datetime.date.today()

	result_object = {
		"year": date.year,
		"month": date.month,
		"day": date.day
	}

	# Можно кешировать на прокси, в течение одного дня
	#   дата точно не поменяется!
	# Очень хорошо про cache-control.
	#   https://stackoverflow.com/a/70970543
	# Создано сегодня в нашем часовом поясе. Переводим после
	#   этого в UTC, обычно в http передают utc время..
	created_datetime = datetime.datetime(date.year, date.month, date.day).astimezone(datetime.UTC)
	HTTP_TIME_FORMAT = "%a, %d %b %Y %H:%M:%S %Z"
	http_time = created_datetime.strftime(HTTP_TIME_FORMAT)
	headers = {
		"Last-Modified": http_time,
		"Date": http_time,
		"Cache-Control": "public, max-age=%d" % (24 * 60 * 60)
	}

	response = flask.jsonify([result_object] * 10_000)
	for header_name, header_value in headers.items():
		response.headers[header_name] = header_value
	return response


@app.route("/what-is-my-name", methods=["POST"])
def nameHandler():
	name = flask.request.form.get("name", "")

	result_object = {
		"name": name
	}

	response = flask.jsonify([result_object] * 10_000)
	return response

File: benchmark-2/backend/backend.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"runtime"
	"time"
)

type date_api_response struct {
	Year  int `json:"year"`
	Month int `json:"month"`
	Day   int `json:"day"`
}

type name_api_response struct {
	Name string `json:"name"`
}

func repeatedSlice[T any](value T, n int) []T {
	arr := make([]T, n)
	for i := 0; i < n; i++ {
		arr[i] = value
	}
	return arr
}

func dateHandler(wr http.ResponseWriter, req *http.Request) {
	if req.Method != "GET" {
		wr.WriteHeader(http.StatusMethodNotAllowed)
		return
	}

	year, month, day := time.Now().Date()
	cur_date := date_api_response{
		Year:  year,
		Month: int(month),
		Day:   day,
	}

	// Можно кешировать на прокси, в течение одного дня
	//   дата точно не поменяется!
	// Очень хорошо про cache-control.
	//   https://stackoverflow.com/a/70970543
	time_generated_str := time.Date(year, month, day, 00, 00, 00, 00, time.Now().Location()).UTC().Format(http.TimeFormat)
	wr.Header().Set("Last-Modified", time_generated_str)
	wr.Header().Set("Date", time_generated_str)
	wr.Header().Set("Cache-Control", fmt.Sprintf("public, max-age=%d", 24*60*60))

	dates := repeatedSlice(cur_date, 10_000)
	json.NewEncoder(wr).Encode(dates)
}

func nameHandler(wr http.ResponseWriter, req *http.Request) {
	if req.Method != "GET" {
		wr.WriteHeader(http.StatusMethodNotAllowed)
		return
	}

	var name_struct name_api_response
	err := json.NewDecoder(req.Body).Decode(&name_struct)
	if err != nil {
		wr.WriteHeader(http.StatusBadRequest)
		wr.Write([]byte("400 - Bad Request (invalid POST parameters)"))
		return
	}

	names := repeatedSlice(name_struct, 10_000)
	json.NewEncoder(wr).Encode(names)
}

func main() {
	// The reason I'm locking goroutines to a single thread
	//   (like limiting their threadpool to a single OS thread)
	//   is because I won't be able to see improvements with
	//   nginx balancing otherwise. Golang will utilize all
	//   cores for request processing. I will have to nginx proxy
	//   between machines to really see benefits of balancing.
	//   So let's make each backend instance seem like a
	//   low-performance machine.
	// Outcome: didn't make difference.
	// runtime.LockOSThread()
	// https://www.reddit.com/r/golang/comments/bj2zq5/comment/em4xgzl
	// Also, I am able to see difference with caching. So that's
	//   great I'll also make another backend, which will be slower.
	// To see the difference I may need to have more resource-heavy
	//   operations in handlers. Then I'll see balancing improving
	//   things.

	// Скажем, не все ресурсы доступны, а живем на отдельной машине с 3 ядрами.
	//   Тогда балансировка позволит включить в работу 6 ядер! А не три.
	runtime.GOMAXPROCS(1)

	http.HandleFunc("/what-date-is-it", dateHandler)
	http.HandleFunc("/what-is-my-name", nameHandler)

	http.ListenAndServe("0.0.0.0:80", nil)
}
